### PR TITLE
[Beam Search] Correct returned beam scores

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

**🚨 🚨 🚨 BREAKING CHANGE: This PR is a breaking change to the values of the generate output `scores` 🚨 🚨 🚨**

It concerns the following use cases:

```python
num_beams =  # > 2 enabling beam search, beam sample or group beam search
generate_outputs = generate(...., num_beams=num_beams, return_dict_in_output=True, output_scores=True)
scores = generate_outputs.scores
```

After multiple issues and discussions, which are best summarized here:
- https://discuss.huggingface.co/t/generation-probabilities-how-to-compute-probabilities-of-output-scores-for-gpt2/3175/15?u=patrickvonplaten
- https://github.com/huggingface/transformers/issues/14612
- https://github.com/huggingface/transformers/issues/14065
- https://github.com/huggingface/transformers/issues/14086

it seems like the best solution here is to do a breaking change in that the output beam scores of generate now correspond better to the transition scores.

The added tests to this PR should explain nicely how one can retrieve all the relevant information from the beam scores in combination with newly added beam_indices.

Very curious to also hear your thoughts on this PR:
- @felix-schneider
- @hacobe
- @qqaatw

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/generation_utils.py [hunk 1]
```diff
@@ -208,10 +208,13 @@ class BeamSearchDecoderOnlyOutput(ModelOutput):
         sequences_scores (`torch.FloatTensor` of shape `(batch_size*num_return_sequences)`, *optional*, returned when `output_scores=True` is passed or when `config.output_scores=True`):
             Final beam scores of the generated `sequences`.
         scores (`tuple(torch.FloatTensor)` *optional*, returned when `output_scores=True` is passed or when `config.output_scores=True`):
-            Processed beam scores for each vocabulary token at each generation step. Beam scores consisting of log
-            softmax scores for each vocabulary token and sum of log softmax of previously generated tokens in this beam
-            . `(max_length-input_ids.shape[-1],)`-shaped tuple of `torch.FloatTensor` with each tensor of shape
+            Beam transition scores for each vocabulary token at each generation step. Beam transition scores consisting
+            of log probabilities of tokens conditioned on log softmax of previously generated tokens in this beam. .
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:212]:
```suggestion
            of log probabilities of tokens conditioned on log softmax of previously generated tokens in this beam.
```

File: src/transformers/generation_utils.py [hunk 2]
```diff
@@ -208,10 +208,13 @@ class BeamSearchDecoderOnlyOutput(ModelOutput):
...
+            `(max_length-input_ids.shape[-1],)`-shaped tuple of `torch.FloatTensor` with each tensor of shape
             `(batch_size*num_beams*num_return_sequences, config.vocab_size)`).
+        beam_indices (`tuple(tuple(torch.LongTensor))`, *optional*, returned when `output_scores=True` is passed or when `config.output_scores=True`):
+            Beam indices of generated token id at each generation step. `(batch_size*num_return_sequences)`-shaped
+            tuple of `(max_length-input_ids.shape[-1],)-shaped tuples of scalar `torch.LongTensor` tensors.
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:217]:
```suggestion
            tuple of `(max_length-input_ids.shape[-1],)`-shaped tuples of scalar `torch.LongTensor` tensors.
```

File: src/transformers/generation_utils.py [hunk 3]
```diff
@@ -241,10 +245,13 @@ class BeamSearchEncoderDecoderOutput(ModelOutput):
         sequences_scores (`torch.FloatTensor` of shape `(batch_size*num_return_sequences)`, *optional*, returned when `output_scores=True` is passed or when `config.output_scores=True`):
             Final beam scores of the generated `sequences`.
         scores (`tuple(torch.FloatTensor)` *optional*, returned when `output_scores=True` is passed or when `config.output_scores=True`):
-            Processed beam scores for each vocabulary token at each generation step. Beam scores consisting of log
-            softmax scores for each vocabulary token and sum of log softmax of previously generated tokens in this beam
-            . `(max_length-1,)`-shaped tuple of `torch.FloatTensor` with each tensor of shape `(batch_size*num_beams,
+            Beam transition scores for each vocabulary token at each generation step. Beam transition scores consisting
+            of log probabilities of tokens conditioned on log softmax of previously generated tokens in this beam. .
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:249]:
```suggestion
            of log probabilities of tokens conditioned on log softmax of previously generated tokens in this beam.
```

File: src/transformers/generation_utils.py [hunk 4]
```diff
@@ -241,10 +245,13 @@ class BeamSearchEncoderDecoderOutput(ModelOutput):
...
+            `(max_length-1,)`-shaped tuple of `torch.FloatTensor` with each tensor of shape `(batch_size*num_beams,
             config.vocab_size)`).
+        beam_indices (`tuple(tuple(torch.LongTensor))`, *optional*, returned when `output_scores=True` is passed or when `config.output_scores=True`):
+            Beam indices of generated token id at each generation step. `(batch_size*num_return_sequences)`-shaped
+            tuple of `(max_length-1,)-shaped tuples of scalar `torch.LongTensor` tensors.
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:254]:
```suggestion
            tuple of `(max_length-1,)`-shaped tuples of scalar `torch.LongTensor` tensors.
```

File: src/transformers/generation_utils.py [hunk 5]
```diff
@@ -286,10 +294,13 @@ class BeamSampleDecoderOnlyOutput(ModelOutput):
         sequences_scores (`torch.FloatTensor` of shape `(batch_size * num_return_sequence)`, *optional*, returned when `output_scores=True` is passed or when `config.output_scores=True`):
             Final beam scores of the generated `sequences`.
         scores (`tuple(torch.FloatTensor)` *optional*, returned when `output_scores=True` is passed or when `config.output_scores=True`):
-            Processed beam scores for each vocabulary token at each generation step. Beam scores consisting of log
-            softmax scores for each vocabulary token and sum of log softmax of previously generated tokens in this beam
-            . `(max_length-input_ids.shape[-1],)`-shaped tuple of `torch.FloatTensor` with each tensor of shape
+            Beam transition scores for each vocabulary token at each generation step. Beam transition scores consisting
+            of log probabilities of tokens conditioned on log softmax of previously generated tokens in this beam. .
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:298]:
```suggestion
            of log probabilities of tokens conditioned on log softmax of previously generated tokens in this beam.
```

File: src/transformers/generation_utils.py [hunk 6]
```diff
@@ -286,10 +294,13 @@ class BeamSampleDecoderOnlyOutput(ModelOutput):
...
+            `(max_length-input_ids.shape[-1],)`-shaped tuple of `torch.FloatTensor` with each tensor of shape
             `(batch_size*num_beams*num_return_sequences, config.vocab_size)`).
+        beam_indices (`tuple(tuple(torch.LongTensor))`, *optional*, returned when `output_scores=True` is passed or when `config.output_scores=True`):
+            Beam indices of generated token id at each generation step. `(batch_size*num_return_sequences)`-shaped
+            tuple of `(max_length-input_ids.shape[-1],)-shaped tuples of scalar `torch.LongTensor` tensors.
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:303]:
```suggestion
            tuple of `(max_length-input_ids.shape[-1],)`-shaped tuples of scalar `torch.LongTensor` tensors.
```

File: src/transformers/generation_utils.py [hunk 7]
```diff
@@ -319,10 +331,13 @@ class BeamSampleEncoderDecoderOutput(ModelOutput):
         sequences_scores (`torch.FloatTensor` of shape `(batch_size * num_return_sequence)`, *optional*, returned when `output_scores=True` is passed or when `config.output_scores=True`):
             Final beam scores of the generated `sequences`.
         scores (`tuple(torch.FloatTensor)` *optional*, returned when `output_scores=True` is passed or when `config.output_scores=True`):
-            Processed beam scores for each vocabulary token at each generation step. Beam scores consisting of log
-            softmax scores for each vocabulary token and sum of log softmax of previously generated tokens in this beam
-            . `(max_length-1,)`-shaped tuple of `torch.FloatTensor` with each tensor of shape `(batch_size*num_beams,
+            Beam transition scores for each vocabulary token at each generation step. Beam transition scores consisting
+            of log probabilities of tokens conditioned on log softmax of previously generated tokens in this beam. .
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:335]:
```suggestion
            of log probabilities of tokens conditioned on log softmax of previously generated tokens in this beam.
```

File: src/transformers/generation_utils.py [hunk 8]
```diff
@@ -319,10 +331,13 @@ class BeamSampleEncoderDecoderOutput(ModelOutput):
...
+            `(max_length-1,)`-shaped tuple of `torch.FloatTensor` with each tensor of shape `(batch_size*num_beams,
             config.vocab_size)`).
+        beam_indices (`tuple(tuple(torch.LongTensor))`, *optional*, returned when `output_scores=True` is passed or when `config.output_scores=True`):
+            Beam indices of generated token id at each generation step. `(batch_size*num_return_sequences)`-shaped
+            tuple of `(max_length-1,)-shaped tuples of scalar `torch.LongTensor` tensors.
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:340]:
```suggestion
            tuple of `(max_length-1,)`-shaped tuples of scalar `torch.LongTensor` tensors.
```

File: tests/test_generation_utils.py [hunk 9]
```diff
@@ -1903,3 +1903,122 @@ def test_generate_encoder_outputs_attention_mask(self):
...
+        scores = torch.stack(outputs.scores).reshape(len(outputs.scores), -1).transpose(0, 1)
+        cut_idx = outputs.sequences.shape[-1] - scores.shape[-1]
+        # compute real indices
+        indices = (
+            outputs.sequences[:, cut_idx:]
+            + torch.tensor(outputs.beam_indices, device=outputs.sequences.device) * vocab_size
+        )
+
+        # gather scores and run
+        return scores.gather(0, indices).sum(-1)
```
Comment (Narsil) [inline tests/test_generation_utils.py:1919]: This seems like a core algorithm so IMHO it could use some code clarification.
Also in that way you don't need to pass `vocab_size` externally.
And that function seems really useful (without the final sum).

Couldn't we instead make it part of `BeamDecoderOutput` or a raw function in `generation_utils` ?
Seems to me at least that the returned tensor is what I would expect roughly from `outputs.scores` (maybe `outputs.generated_scores()` to not break BC ?)

```suggestion
    def compute_transition_beam_scores_sum(outputs):

        # reshape scores
        # output.scores = [seq_len, batch_size, vocab_size]
        scores = torch.stack(outputs.scores)
        (seq_len, batch_size, vocab_size) = scores.shape
        scores = scores.reshape(seq_len, -1).transpose(0, 1)
        # scores = [vocab_size * batch_size, seq_len]

        # The output sequences might contain some decoder_initial_ids. Ignore those.
        cut_idx = outputs.sequences.shape[-1] - scores.shape[-1]

        # compute real indices
        beam_offsets = torch.tensor(outputs.beam_indices, device=outputs.sequences.device) * vocab_size
        indices = outputs.sequences[:, cut_idx:] + beam_offsets

        # gather scores and run
        return scores.gather(0, indices).sum(-1)
```

**Review by Narsil — COMMENTED** [review]
Contrary, to the initial PR comment, I don't see any breaking changes.
I only see the addition of `beam_indices` which seem to enable calculating the transition scores from `outputs.scores`. Is that right ?

- Is it breaking and we didn't have a test before that had to be modified ? (I didn't see any code either that made it look like scores had changed)

- If it is not breaking, it seems the test function `compute_scores` is very interesting to users too and is what was asked for, did I understand that correctly ?
- If yes, maybe we should move that to library code ?

**Comment by Narsil** [comment]
Other related PR maybe : https://github.com/huggingface/transformers/issues/14921

File: tests/test_generation_utils.py (hunk 9 above)
Comment (felix-schneider) [reply tests/test_generation_utils.py:1919]: This (without the sum) is much more likely to be what a user wants than `scores`. Imho, it should be included with the Beam Search Output rather than `scores` and `beam_indices`, both of which are beam search internal values, whereas the per-token output probabilities are not.

I would not include the sum in this function, as the user already gets the sequence score as a separate output. The reason why the user is giving `return_scores=True` is because they want per-token logprobs.

How would this function (and `sequences_scores`) deal with return sequences of unequal length? Is the output after EOS filled with pad and logprob `-inf`? Or could there be garbage output and arbitrary probabilities?

File: tests/test_generation_utils.py [hunk 10]
```diff
@@ -1903,3 +1903,122 @@ def test_generate_encoder_outputs_attention_mask(self):
...
+            length_penalty=0.0,
+        )
+        model = model.to(torch_device)
+
+        input_ids = tokenizer(articles, return_tensors="pt", padding=True).input_ids.to(torch_device)
+        outputs = model.generate(input_ids=input_ids)
+
+        transition_scores_sum = self.compute_transition_beam_scores_sum(outputs, model.config.vocab_size)
+
+        self.assertTrue(torch.allclose(transition_scores_sum, outputs.sequences_scores, atol=1e-3))
```
Comment (felix-schneider) [inline tests/test_generation_utils.py:1944]: Very important test, I'm glad it was included. The workaround I posted in the original issue did not give the exact same scores as `sequences_scores`, so it's good that this approach does.

File: src/transformers/generation_utils.py [hunk 11]
```diff
@@ -319,10 +331,13 @@ class BeamSampleEncoderDecoderOutput(ModelOutput):
         sequences_scores (`torch.FloatTensor` of shape `(batch_size * num_return_sequence)`, *optional*, returned when `output_scores=True` is passed or when `config.output_scores=True`):
             Final beam scores of the generated `sequences`.
         scores (`tuple(torch.FloatTensor)` *optional*, returned when `output_scores=True` is passed or when `config.output_scores=True`):
-            Processed beam scores for each vocabulary token at each generation step. Beam scores consisting of log
-            softmax scores for each vocabulary token and sum of log softmax of previously generated tokens in this beam
-            . `(max_length-1,)`-shaped tuple of `torch.FloatTensor` with each tensor of shape `(batch_size*num_beams,
+            Beam transition scores for each vocabulary token at each generation step. Beam transition scores consisting
+            of log probabilities of tokens conditioned on log softmax of previously generated tokens in this beam.
+            `(max_length-1,)`-shaped tuple of `torch.FloatTensor` with each tensor of shape `(batch_size*num_beams,
             config.vocab_size)`).
```
Comment (felix-schneider) [inline src/transformers/generation_utils.py:337]: Good change clarifying "softmax scores" to "log probabilities". But the doc still mentions softmax just after the changed part:

```suggestion
            Beam transition scores for each vocabulary token at each generation step. Beam transition scores consisting
            of log probabilities of tokens plus log probabilities of previously generated tokens in this beam.
            `(max_length-1,)`-shaped tuple of `torch.FloatTensor` with each tensor of shape `(batch_size*num_beams,
            config.vocab_size)`).
```

**Review by felix-schneider — COMMENTED** [review]
I think adding another output will just make the output more confusing. The current `scores` output still does not add anything of value imo and should be replaced with the output of `compute_transition_beam_scores_sum` without the sum.

Users that are interested in beam search internals such as `beam_indices` and the transition scores can still roll their own beam search to output these internals but the majority of users of `return_scores=True` will be interested in the per-token logprobs.

**Review by sgugger — COMMENTED** [review]
I'll trust your judgement on this one @patrickvonplaten, but note that we can also add more fields to the generation outputs when `return_scores=True`.

**Comment by patrickvonplaten** [comment]
Thanks a lot for your feedback here guys and sorry to only pick up the PR again.

@Narsil @felix-schneider,

I see that you guys would like to make it easier for the user to directly get the transition probabilities of the generated tokens. 

First, let's clarify a couple of things:
@Narsil, this PR is breaking in that the `scores` output is now different in terms of it's values (the shape stays the same). We have only ever tested for the shape to be correct, but never whether the output values or "correct" (which is a bit weird as it would be similar to testing whether the values of the attention matrices correspond to some hardcoded values. Nevertheless, we verify this now in the test added here. 

So this is indeed a big breaking change (it's somewhat equal to changing the output values of the attention matrices when setting `output_attentions=True`), but given that `scores` output has been an experimental feature and is quite complex for generation, I think we can make an exception here, given all the issues above. Also cc @LysandreJik @sgugger @patil-suraj here - no need to understand the change in-detail, but maybe just so that you're aware that this PR is a big breaking change for people doing `model.generate(...., return_dict_in_generate=True, output_scores=True)`.

Second, @felix-schneider, I do understand your point of view in that outputting the raw `scores` doesn't add much value. However, it is important to keep the following in mind:

1. On of the philosophies of `transformers` is the following:
```
Expose the models’ internals as consistently as possible:

We give access, using a single API, to the full hidden-states and attention weights.
Tokenizer and base model’s API are standardized to easily switch between mode
``` 
(see https://huggingface.co/docs/transformers/philosophy)

This does very well apply here as well. We are simply outputting the raw scores that are used to sample the next token in generation which is in my opinion quite consistent with "expose the models' internals". 

2. The library is heavily used by researchers who do whatsoever with the model internals for analysis, complex algorithms etc...in my experience, there are always a lot of use cases that one doesn't take into account when trying to design an API that "makes most sense for the end user" meaning that often one thinks there is only one use case and that only this use case should be implemented. 
So far the approach of exposing the model's internals worked well IMO (e.g. many people like the fact that the attention mask is that easily available), but as you pointed out it does have the downside to make it more complex and potentially more difficult (especially for development/production oriented users). So it's always a trade-off when deciding what values should be exposed and how.

3. It's **very very** important that we keep backward compatibility. While this PR already breaks Backward comp in some sense, we should definitely not fully remove the `scores` output - all users that have used `scores` before would then have no way to use `transformers` for their use case anymore -> so this cannot be done.

4. We have to be careful to not make maintenance explode here. The less "bare bone" the output is, *i.e.* the more specific, the more prone it is to have errors and the costlier is the maintenance


As a conclusion, I do like @Narsil idea of adding the functionality of computing the transition probabilites as a function to the output classes (or to `generate()` itself). This way, it's very easy for the user to compute the transition scores (just pass `scores` and `beam_indices` into `model.compute_transition_probabilities(scores, beam_indices)`) this way adding a single line would be enough for the user. 
I agree with @felix-schneider that we should not sum up the probs and we should probably also allow the user to pass an `attention_mask` so that the final values are correctly masked. 

I will adapt the PR according to this. It would be great if you could then take a second look whether it fits your needs

File: src/transformers/generation_utils.py [hunk 12]
```diff
@@ -2231,14 +2308,14 @@ def beam_sample(
...
-            next_token_scores = next_token_scores + beam_scores[:, None].expand_as(next_token_scores)
+            next_token_scores_processed = logits_processor(input_ids, next_token_scores)
+            next_token_scores = next_token_scores_processed + beam_scores[:, None].expand_as(next_token_scores)
             next_token_scores = logits_warper(input_ids, next_token_scores)
 
             # Store scores, attentions and hidden_states when required
             if return_dict_in_generate:
                 if output_scores:
-                    scores += (next_token_scores,)
+                    scores += (logits_warper(input_ids, next_token_scores_processed),)
```
Comment (Narsil) [inline src/transformers/generation_utils.py:2318]: This is the breaking change then ?

This is what I missed earlier I guess.

File: src/transformers/generation_utils.py [hunk 13]
```diff
@@ -743,6 +759,47 @@ def _merge_criteria_processor_list(
         default_list.extend(custom_list)
         return default_list
 
+    def compute_transition_beam_scores(
+        self,
+        sequences: torch.Tensor,
+        scores: Tuple[torch.Tensor],
+        beam_indices: torch.Tensor,
+        eos_token_id: int = None,
+        dummy=None,
```
Comment (Narsil) [inline src/transformers/generation_utils.py:768]: Is that necessary ?

File: src/transformers/generation_utils.py [hunk 14]
```diff
@@ -743,6 +759,47 @@ def _merge_criteria_processor_list(
...
+    ):
+        """compute the transition probabilities of sequences given generation
+        scores and beam indices"""
+        # reshape scores
+        scores = torch.stack(scores).reshape(len(scores), -1).transpose(0, 1)
```
Comment (Narsil) [inline src/transformers/generation_utils.py:773]: I really like comments explaining the expected shapes of tensors in here to make the code more readable/understandable.

`#reshape scores` is not very informative (since I can read it in the following line, along with `stack` and `transpose`.
What I am missing as a reader is:
- Why you are doing this?
- What should the final result be ?

`# Reshape scores as [vocab_size * batch_size, seq_len] in order to use a single gather op later` is a suggestions

**Review by Narsil — APPROVED** [review]
LGTM overall with a few comments.

Regarding tests with values, I imagine having a handful could be interesting to make sure we're not breaking compatibility later inadvertently.

**Comment by felix-schneider** [comment]
I see your point, @patrickvonplaten. My suggestion though would be to have the `compute_transition_probabilities` method live with the object returned by `generate`, rather than the model, as this object already has access to all required values: `scores`, `beam_indices` and `attention_mask`.

Putting it with the model implies the ability for different models to modify this behaviour. I'm not sure that this is actually possible as getting transition scores from model outputs is done by beam search and not customizable by models.

File: src/transformers/generation_utils.py (hunk 13 above)
Comment (patrickvonplaten) [reply src/transformers/generation_utils.py:768]: great catch! No this was for debugging haha

**Comment by patrickvonplaten** [comment]
> I see your point, @patrickvonplaten. My suggestion though would be to have the `compute_transition_probabilities` method live with the object returned by `generate`, rather than the model, as this object already has access to all required values: `scores`, `beam_indices` and `attention_mask`.
> 
> Putting it with the model implies the ability for different models to modify this behaviour. I'm not sure that this is actually possible as getting transition scores from model outputs is done by beam search and not customizable by models.

I see your point! We could add the function (also as @Narsil suggested) to the `BeamSearchEncoderDecoderOutput` class as you rightly say it's pretty "generation function"-specific. I'm still leaning against it though because, we've never done this kind of design in `transformers` and I'm not sure it's very intuitive for our users as the workflow would look as follows:

```python 
sequence_outputs = model.generate(..., output_scores=True, return_dict_in_generate=True)
transition_scores = sequence_outputs.compute_transition_scores(vocab_size=model.config.vocab_size)   # not that it's also a a bit tedious to get the vocab size here (we might though actually retrieve that from the shape of some outputs (not trivial though)
```

I think there are three possible designs we could do here:

- 1. The one shown above
- 2. What's implemented now:
```python
sequence_outputs = model.generate(..., output_scores=True, return_dict_in_generate=True)
transition_scores = model.compute_transition_scores(sequence_outputs.sequences, sequence_outputs.scores, sequence_outputs.beam_indices)
```
- 3. Just compute it when scores should be returned:
```python
sequence_outputs = model.generate(..., output_scores=True, return_dict_in_generate=True)
sequence_outputs.transition_scores
```

While 3 is the nicest for the end-user, it is a **very** specific (it creates the transition scores of only the generated sequence ids). Also, it creates quite some maintenance which I would like to avoid a bit.

To summarize, I think 2 is the best way to go because it's general in a sense that people can create the transition probs of any possible sequence of ids (not just the generated ones). 

Keen to hear your final thoughts

**Comment by Narsil** [comment]
I agree with `2.` It's easier to upgrade later, than downgrade `3.` later (if maintenance becomes hard).

**Comment by qqaatw** [comment]
I like the `3.` more than `2.` if the maintenance cost would not become too heavy as IMO it looks more intuitive to users to get the transition scores that most users are looking for (I think). We could also keep `model.compute_transition_scores` in case users want to calculate their own scores.

**Comment by patrickvonplaten** [comment]
Ok seems like we need some votes here for 2. vs 3. haha. 

@patil-suraj @LysandreJik @sgugger what do you think?

File: src/transformers/generation_utils.py [hunk 15]
```diff
@@ -743,6 +759,49 @@ def _merge_criteria_processor_list(
...
+        scores and beam indices"""
+
+        # reshape scores as [vocab_size * batch_size, # generation steps]
+        # with batch_size being 2 * vocab_size and # generation steps being
+        # seq_len - input_length
+        scores = torch.stack(scores).reshape(len(scores), -1).transpose(0, 1)
+
+        # start of generated tokens
+        cut_idx = sequences.shape[-1] - scores.shape[-1]
+
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:779]:
```suggestion
```

File: src/transformers/generation_utils.py [hunk 16]
```diff
@@ -743,6 +759,49 @@ def _merge_criteria_processor_list(
...
+        # adjust for beam indices
+        beam_sequence_indices = torch.tensor(beam_indices, device=sequences.device) * self.config.vocab_size
+
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:782]:
```suggestion
```

File: src/transformers/generation_utils.py [hunk 17]
```diff
@@ -743,6 +759,49 @@ def _merge_criteria_processor_list(
...
+        # compute real indices
+        indices = sequences[:, cut_idx:] + beam_sequence_indices
+
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:785]:
```suggestion
```

File: src/transformers/generation_utils.py [hunk 18]
```diff
@@ -743,6 +759,49 @@ def _merge_criteria_processor_list(
...
+        # gather scores and run
+        transition_scores = scores.gather(0, indices)
+
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:788]:
```suggestion
```

**Comment by sgugger** [comment]
I don't have enough context in terms of what "the burden of maintenance" entails for option 3 there, but it would be the most elegant in my opinion. But as a propriety to only be computed when asked for, not as a regular attribute that would be computed all the time (if that's possible, if not we should fall back to option 2).

I'm also okay with option 2.

**Comment by patrickvonplaten** [comment]
Taking everything into account, I'm going for option 2. now (which is the one implemented here). Given that:
    - we can always go from 2. to 3. (but not the other way around because of backward comp), 
    - 2. is more general than 3. at lower maintenance - I see many use cases of computing the scores of not generated ids, but almost generated ids
    - and the fact that this is a very researchy feature in general where some knowledge about `generate` can be expected

2 is the better option here. Let's see in the coming month how much this feature is used. I'll also try to make an updated blog post about `generate` in general where I'll include this feature

**Comment by rafikg** [comment]
hi @patrickvonplaten,
It seems that your PR is not available yet in transformers =4.18.0 . My first attempt as many people :)  is to try to match between the `scores` and the sequnces using softmax until I see your discussion.

**Comment by yuezih** [comment]
Hi @patrickvonplaten,
Been trying to replicate `sequence_outputs.sequences` through `sequence_outputs.scores` and `sequences_outputs.beam_indices` until I saw your discussion, and glad to have found resonance. 
But this still doesn't solve my confusion. What if I want to get not only beam scores, but the whole probability distribution on the vocab? (That means a return of `torch.Size([batch_size, seq_len, vocab_size])` from `model.compute_transition_scores()`)
